src/vec.rs

```diff
@@ -0,0 +1,637 @@
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+
+use core::marker::PhantomData;
+use core::ops::{Deref, DerefMut};
+use core::convert::{AsRef, AsMut};
+use core::ptr;
+use core::cmp;
+
+/// A continuous, growable array type
+pub struct Vec<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    _marker: PhantomData<[T]>,
+    array: A,
+    len: usize,
+}
+
+impl<T, A> Vec<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    /// Creates a new vector using `array` as the backup storage
+    ///
+    /// # Example
+    /// ```
+    /// #![feature(const_fn)]
+    /// use heapless::Vec;
+    /// const LEN: usize = 16;
+    /// // with let
+    /// let vec_let = Vec::new([0;LEN]);
+    /// // with const
+    /// const VEC_CONST: Vec<u8, [u8; LEN]> = Vec::new([0;LEN]);
+    /// // with static
+    /// static VEC: Vec<u8, [u8; LEN]> = Vec::new([0;LEN]);
+    /// // with slice
+    /// let mut array = [0_usize; 8];
+    /// let mut vec_slice = Vec::new(&mut array);
+    /// ```
+    #[inline]
+    pub const fn new(array: A) -> Self {
+        unsafe { Self::with_len(array, 0) }
+    }
+
+    /// Creates a new vector using `array` as the backup storage.
+    /// Keeps the values and length of the array.
+    /// The backup storage is used up to the provided length upon creation.
+    /// This function can be used to quickly initialize the vector.
+    ///
+    /// The initial length must be passed,
+    /// because the length of the array might not be determined
+    /// at compile time.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::Vec;
+    /// let vec = unsafe{ Vec::with_len([1,2,3], 3) };
+    /// assert!(vec.is_full());
+    /// ```
+    #[inline]
+    pub const unsafe fn with_len(array: A, len: usize) -> Self {
```

Comment:
This seems rather easy to screw up. Could this become safe by adding an assertion / check into the implementation?

Reply:
We would need some kind of compile time assertions here. I don't know if this is in the compiler yet or if there are any crates providing this. Furthermore the length of the array is not known on compile time. Maybe we should drop this function for now.